### ISSUE TITLE
PyTorch model slows down 30x when binding threads with fewer than half the CPU cores.

### ISSUE THREAD

**Issue by ye3084** [pytorch/pytorch#125547]
### 🐛 Describe the bug

When I run VGG19 100 times with more than 8 threads, it takes 6 seconds, but when I reduce the number of threads to 7, it takes 184 seconds. This result confuses me. Is it a problem with my code or a bug?
I'm using a device with a single user, equipped with a 16-core CPU. 
I've tested TensorFlow and TVM but did not encounter this issue.
`import threading
import os
import time
import torch
from torchvision.models import vgg19

def run_vgg_model(core_ids, epoch=100):
    current = threading.current_thread()
    os.sched_setaffinity(current.native_id, core_ids)

    print(f"thread start, CPU cores：{os.sched_getaffinity(0)}\n", end='')

    model = vgg19(weights=None)
    model.eval()

    input_tensor = torch.rand(1, 3, 224, 224)
    start_time = time.time()
    for i in range(epoch):
        model(input_tensor)
    end_time = time.time()

    thread_use_time = end_time - start_time

    print(f"thread_use_time：{thread_use_time:.5f}")

if __name__=="__main__":
    NUM_CORES = os.cpu_count()
    core_ids = list(range(NUM_CORES))
    core_ids1 = core_ids[:NUM_CORES // 2]
    core_ids2 = core_ids[:NUM_CORES // 2 - 1]

    # thread = threading.Thread(target=run_vgg_model, args=(core_ids,))
    # thread = threading.Thread(target=run_vgg_model, args=(core_ids1,))
    thread = threading.Thread(target=run_vgg_model, args=(core_ids2,))
    thread.start()`
<img width="497" alt="图片1" src="https://github.com/pytorch/pytorch/assets/115478284/8185bb18-0d9f-45a2-8601-ddbba1c9b698">
<img width="307" alt="图片2" src="https://github.com/pytorch/pytorch/assets/115478284/328c71e7-8368-440a-88ea-aa6c2b0ffd04">
<img width="298" alt="图片3" src="https://github.com/pytorch/pytorch/assets/115478284/94f17144-681e-4d26-a62c-334eaeada841">

### Versions

Collecting environment information...
PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.3
Libc version: glibc-2.31

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-177-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             16
On-line CPU(s) list:                0-15
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) W-2145 CPU @ 3.70GHz
Stepping:                           4
CPU MHz:                            1200.202
CPU max MHz:                        4500.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           7399.70
Virtualization:                     VT-x
L1d cache:                          256 KiB
L1i cache:                          256 KiB
L2 cache:                           8 MiB
L3 cache:                           11 MiB
NUMA node0 CPU(s):                  0-15
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: Split huge pages
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT vulnerable
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[pip3] torchvision==0.18.0
[pip3] triton==2.3.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] torchvision               0.18.0                   pypi_0    pypi
[conda] triton                    2.3.0                    pypi_0    pypi

**Comment by malfet** [comment]
It feels like this issue belong to https://discuss.pytorch.org/

Though if I'm to rephrase it, it probably about documenting what are the interoperability expectations between PyTorch runtime (which uses OpenMP) and Python threading library

**Comment by malfet** [comment]
Adding `triage review` to discuss what do do about various performance issue and how we should approach them

**Comment by tringwald** [comment]
By using `.sched_setaffinity()` you are basically telling your OS to only allow the current process to use certain CPU cores. `torch` will use `torch.get_num_threads()` (usually CPU cores/2) threads for computation by default. So in your example you will force the OS to schedule N threads on N-1 cores which probably leads to some very heavy congestion. 
You should probably use `torch.set_num_threads()` to limit the number of threads `torch` will use.

**Comment by ye3084** [comment]
> By using `.sched_setaffinity()` you are basically telling your OS to only allow the current process to use certain CPU cores. `torch` will use `torch.get_num_threads()` (usually CPU cores/2) threads for computation by default. So in your example you will force the OS to schedule N threads on N-1 cores which probably leads to some very heavy congestion. You should probably use `torch.set_num_threads()` to limit the number of threads `torch` will use.

Thank you for your response. I used torch.set_num_threads() and it solved the problem, but I found that when I use more than 8 cores, the thread execution time increases. It seems that using 8 cores gives the best performance. I've created a chart where the X-axis represents the number of cores used and the Y-axis shows the execution time for the threads.
![img](https://github.com/pytorch/pytorch/assets/115478284/88cc3c36-7907-493e-bc48-caf1e4b6af47)

The results I obtained on another server with a 72-core CPU are as follows (this server is shared with other users, which might affect performance).
![img2](https://github.com/pytorch/pytorch/assets/115478284/322670f8-f87f-4a57-9d0a-150941d0f545)

**Comment by ye3084** [comment]
> It feels like this issue belong to https://discuss.pytorch.org/
> 
> Though if I'm to rephrase it, it probably about documenting what are the interoperability expectations between PyTorch runtime (which uses OpenMP) and Python threading library
> 
> Also, I wonder if `torch.set_num_threads(len(core_ids))` will help

Thank you for your response. I used torch.set_num_threads() and it solved the problem, but I found that when I use more than 8 cores, the thread execution time increases. It seems that using 8 cores gives the best performance. I've created a chart where the X-axis represents the number of cores used and the Y-axis shows the execution time for the threads.
![img](https://github.com/pytorch/pytorch/assets/115478284/88cc3c36-7907-493e-bc48-caf1e4b6af47)

The results I obtained on another server with a 72-core CPU are as follows (this server is shared with other users, which might affect performance).
![img2](https://github.com/pytorch/pytorch/assets/115478284/322670f8-f87f-4a57-9d0a-150941d0f545)

**Comment by tringwald** [comment]
Your CPU only has 8 (or 36) physical cores, with hyperthreading/SMT that makes 16/72 logical cores. SMT only really helps when waiting for something like IO. In a heavy number crunching task like NN inference, there is no real advantage of using logical cores. Having multiple threads fight for the same physical core only leads to congestion and cache eviction.
